dandi/cli/cmd_service_scripts.py

```diff
@@ -0,0 +1,103 @@
+    `http` extra (e.g., `pip install "fsspec[http]"`).
+    """
+    from ..metadata import nwb2asset  # Avoid heavy import at top level
+
+    parsed_url = parse_dandi_url(url)
+    if parsed_url.version_id not in (None, "draft"):
+        raise click.UsageError("URL cannot point to a published Dandiset")
+    parsed_url.version_id = "draft"
+    current_schema_version = Version(DANDI_SCHEMA_VERSION)
+    with parsed_url.navigate(authenticate=True, strict=True) as (_, _, assets):
+        for asset in assets:
+            if PurePosixPath(asset.path).suffix.lower() != ".nwb":
+                lgr.info(
+                    "Asset %s (%s) is not NWB; skipping", asset.identifier, asset.path
+                )
+            assert isinstance(asset, RemoteBlobAsset)
+            lgr.info("Processing asset %s (%s)", asset.identifier, asset.path)
+            if when == "always":
+                do_reextract = True
+            else:
+                try:
+                    sv = asset.get_raw_metadata()["schemaVersion"]
+                except KeyError:
+                    do_reextract = True
+                else:
+                    schemaVersion = Version(sv)
+                    if schemaVersion < current_schema_version:
+                        do_reextract = True
+                    elif schemaVersion == current_schema_version:
+                        lgr.info(
+                            "Asset's schemaVersion is up to date; not reextracting"
+                        )
+                        do_reextract = False
+                    else:
+                        lgr.warning(
+                            "schemaVersion of asset %s (%s) is %r, higher than"
+                            " current schema version %r",
+                            asset.identifier,
+                            asset.path,
+                            sv,
+                            DANDI_SCHEMA_VERSION,
+                        )
+                        do_reextract = False
+            if do_reextract:
+                try:
+                    digest = asset.get_digest()
+                except NotFoundError:
+                    digest = None
+                lgr.info("Extracting new metadata for asset")
+                metadata = nwb2asset(asset.as_readable(), digest=digest)
+                metadata.path = asset.path
+                mddict = metadata.json_dict()
+                if diff:
+                    oldmd = asset.get_raw_metadata()
+                    oldmd_str = json.dumps(oldmd, indent=4, sort_keys=True)
+                    mddict_str = json.dumps(mddict, indent=4, sort_keys=True)
+                    print(
+                        unified_diff(
+                            oldmd_str.splitlines(True),
+                            mddict_str.splitlines(True),
+                            fromfile=f"{asset.path}:old",
+                            tofile=f"{asset.path}:new",
+                        )
```

Comment:
needs to be joined or smth 
```
❯ dandi service-scripts reextract-metadata --diff --when newer-schema-version dandi://dandi/000029@draft/sub-anm369963/sub-anm369963_behavior+ecephys.nwb
2023-03-02 17:54:20,940 [    INFO] Processing asset 38f2024d-62a9-4c79-8a22-7a0ff34b331d (sub-anm369963/sub-anm369963_behavior+ecephys.nwb)
2023-03-02 17:54:21,051 [    INFO] Extracting new metadata for asset
/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3.11/lib/python3.11/site-packages/hdmf/spec/namespace.py:531: UserWarning: Ignoring cached namespace 'core' version 2.0.2 because version 2.6.0-alpha is already loaded.
  warn("Ignoring cached namespace '%s' version %s because version %s is already loaded."
<generator object unified_diff at 0x7efd1149a8e0>
2023-03-02 17:54:29,124 [    INFO] Saving new asset metadata
2023-03-02 17:54:29,292 [    INFO] Logs saved in /home/yoh/.cache/dandi-cli/log/20230302225417Z-3416570.log
dandi service-scripts reextract-metadata --diff --when newer-schema-version   6.46s user 2.48s system 65% cpu 13.565 total
```

<details>
<summary>FTR here is the diff which was here (I did manually dumping those yaml structures from `ls`)</summary> 

```diff
--- /tmp/schema-0.4.4.yaml	2023-03-02 17:51:10.904878979 -0500
+++ /tmp/schema-0.6.2.yaml	2023-03-02 17:55:28.152268322 -0500
@@ -1,63 +1,49 @@
-- asset_id: 38f2024d-62a9-4c79-8a22-7a0ff34b331d
+- asset_id: 2afb837e-7480-44d8-a15f-bf3e433a0431
   blob: e7288a2c-444c-42c5-b7bb-9a58c728992b
-  created: '2022-07-21T18:18:44.014978+00:00'
+  created: '2023-03-02T22:54:29.224996+00:00'
   metadata:
-    '@context': https://raw.githubusercontent.com/dandi/schema/master/releases/0.4.4/context.json
+    '@context': https://raw.githubusercontent.com/dandi/schema/master/releases/0.6.3/context.json
     access:
     - schemaKey: AccessRequirements
       status: dandi:OpenAccess
     approach:
-    - name: behavioral approach
-      schemaKey: ApproachType
     - name: electrophysiological approach
       schemaKey: ApproachType
+    - name: behavioral approach
+      schemaKey: ApproachType
     blobDateModified: '2020-12-02T20:33:27.624158-05:00'
     contentSize: 6393196
     contentUrl:
-    - https://api.dandiarchive.org/api/assets/38f2024d-62a9-4c79-8a22-7a0ff34b331d/download/
+    - https://api.dandiarchive.org/api/assets/2afb837e-7480-44d8-a15f-bf3e433a0431/download/
     - https://dandiarchive.s3.amazonaws.com/blobs/e72/88a/e7288a2c-444c-42c5-b7bb-9a58c728992b
-    dateModified: '2021-06-30T17:33:36.230626-04:00'
-    datePublished: '2022-11-07T23:44:44.018303+00:00'
+    dateModified: '2023-03-02T17:54:29.036005-05:00'
     digest:
       dandi:dandi-etag: 54187aaeda80bbb19ed7479c01857548-1
       dandi:sha2-256: ebcdf64353eeb9b605533fdf8ee1c62a710b64e49593fe0f6de2e1ef8715cc7d
     encodingFormat: application/x-nwb
-    id: dandiasset:38f2024d-62a9-4c79-8a22-7a0ff34b331d
-    identifier: 38f2024d-62a9-4c79-8a22-7a0ff34b331d
+    id: dandiasset:2afb837e-7480-44d8-a15f-bf3e433a0431
+    identifier: 2afb837e-7480-44d8-a15f-bf3e433a0431
     keywords:
     - motor planning
     - premotor cortex
     - preparatory activity
     - extracellular electrophysiology
     measurementTechnique:
-    - name: surgical technique
-      schemaKey: MeasurementTechniqueType
     - name: spike sorting technique
       schemaKey: MeasurementTechniqueType
+    - name: surgical technique
+      schemaKey: MeasurementTechniqueType
     - name: behavioral technique
       schemaKey: MeasurementTechniqueType
     path: sub-anm369963/sub-anm369963_behavior+ecephys.nwb
-    publishedBy:
-      endDate: '2022-11-07T23:44:44.018303+00:00'
-      id: urn:uuid:1ada4c80-e22d-43d6-8571-4183a3a2eaa8
-      name: DANDI publish
-      schemaKey: PublishActivity
-      startDate: '2022-11-07T23:44:44.018303+00:00'
-      wasAssociatedWith:
-      - id: urn:uuid:483dfc66-abfc-4b87-b5d8-2a914cc24a77
-        identifier: RRID:SCR_017571
-        name: DANDI API
-        schemaKey: Software
-        version: 0.1.0
-    repository: https://dandiarchive.org/
     schemaKey: Asset
-    schemaVersion: 0.4.4
+    schemaVersion: 0.6.3
     variableMeasured:
     - schemaKey: PropertyValue
-      value: Units
-    - schemaKey: PropertyValue
       value: ElectrodeGroup
     - schemaKey: PropertyValue
+      value: Units
+    - schemaKey: PropertyValue
       value: BehavioralEvents
     wasAttributedTo:
     - genotype: ''
@@ -68,19 +54,25 @@
         schemaKey: SexType
       species:
         identifier: http://purl.obolibrary.org/obo/NCBITaxon_10090
-        name: House mouse
+        name: Mus musculus - House mouse
         schemaKey: SpeciesType
     wasGeneratedBy:
+    - description: ''
+      name: Acquisition session
+      schemaKey: Session
+      startDate: '2017-02-26T00:00:00-06:00'
     - description: Metadata generated by DANDI cli
-      id: urn:uuid:ca8f9621-7834-4a13-bf56-b6833c37a610
+      endDate: '2023-03-02T17:54:29.035866-05:00'
+      id: urn:uuid:8d00e815-4e37-4bb9-9ac7-4c4815f3bb59
       name: Metadata generation
       schemaKey: Activity
+      startDate: '2023-03-02T17:54:21.111054-05:00'
       wasAssociatedWith:
       - identifier: RRID:SCR_019009
         name: DANDI Command Line Interface
         schemaKey: Software
         url: https://github.com/dandi/dandi-cli
-        version: 0.22.0
-  modified: '2022-07-21T18:18:44.068369+00:00'
+        version: 0.50.0+2.gfee08f8c
+  modified: '2023-03-02T22:54:29.259908+00:00'
   path: sub-anm369963/sub-anm369963_behavior+ecephys.nwb
   size: 6393196
```
</details>

Reply:
<details>
<summary>FTR: here is an example of invocation on some other asset now after the fix -- looks good</summary> 

```shell
❯ dandi service-scripts reextract-metadata --diff --when newer-schema-version dandi://dandi/000029@draft/sub-anm369962/sub-anm369962_behavior+ecephys.nwb
2023-03-03 11:14:47,458 [    INFO] Processing asset 4f6a8f1d-ca04-491f-a530-9ed00a909c21 (sub-anm369962/sub-anm369962_behavior+ecephys.nwb)
2023-03-03 11:14:47,504 [    INFO] Asset's schemaVersion '0.4.4' is out of date; will reextract
2023-03-03 11:14:47,553 [    INFO] Extracting new metadata for asset
/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3.11/lib/python3.11/site-packages/hdmf/spec/namespace.py:531: UserWarning: Ignoring cached namespace 'core' version 2.0.2 because version 2.6.0-alpha is already loaded.
  warn("Ignoring cached namespace '%s' version %s because version %s is already loaded."
--- sub-anm369962/sub-anm369962_behavior+ecephys.nwb:old
+++ sub-anm369962/sub-anm369962_behavior+ecephys.nwb:new
@@ -1,4 +1,3 @@
-'@context': https://raw.githubusercontent.com/dandi/schema/master/releases/0.4.4/context.json
 access:
 - schemaKey: AccessRequirements
   status: dandi:OpenAccess
@@ -9,50 +8,30 @@
   schemaKey: ApproachType
 blobDateModified: '2020-12-02T20:33:26.520150-05:00'
 contentSize: 6644036
-contentUrl:
-- https://api.dandiarchive.org/api/assets/4f6a8f1d-ca04-491f-a530-9ed00a909c21/download/
-- https://dandiarchive.s3.amazonaws.com/blobs/180/9f5/1809f541-1cb1-48b8-b916-9a696bab488d
-dateModified: '2021-06-30T17:36:42.319549-04:00'
-datePublished: '2022-11-07T23:44:44.018389+00:00'
+dateModified: '2023-03-03T11:14:56.002044-05:00'
 digest:
   dandi:dandi-etag: 9234087b734d140bed4d65bb83c51c90-1
-  dandi:sha2-256: d01c43576e3aeceed53a052defa8bbbf15b885146f518acf63d228cee44576ac
 encodingFormat: application/x-nwb
-id: dandiasset:4f6a8f1d-ca04-491f-a530-9ed00a909c21
-identifier: 4f6a8f1d-ca04-491f-a530-9ed00a909c21
 keywords:
 - motor planning
 - premotor cortex
 - preparatory activity
 - extracellular electrophysiology
 measurementTechnique:
+- name: spike sorting technique
+  schemaKey: MeasurementTechniqueType
 - name: behavioral technique
-  schemaKey: MeasurementTechniqueType
-- name: spike sorting technique
   schemaKey: MeasurementTechniqueType
 - name: surgical technique
   schemaKey: MeasurementTechniqueType
 path: sub-anm369962/sub-anm369962_behavior+ecephys.nwb
-publishedBy:
-  endDate: '2022-11-07T23:44:44.018389+00:00'
-  id: urn:uuid:808fbb82-aba3-4ce0-805f-cfba8a93e39a
-  name: DANDI publish
-  schemaKey: PublishActivity
-  startDate: '2022-11-07T23:44:44.018389+00:00'
-  wasAssociatedWith:
-  - id: urn:uuid:f4515777-e62e-4eef-a416-7c0ed2543438
-    identifier: RRID:SCR_017571
-    name: DANDI API
-    schemaKey: Software
-    version: 0.1.0
-repository: https://dandiarchive.org/
 schemaKey: Asset
-schemaVersion: 0.4.4
+schemaVersion: 0.6.3
 variableMeasured:
 - schemaKey: PropertyValue
+  value: BehavioralEvents
+- schemaKey: PropertyValue
   value: ElectrodeGroup
-- schemaKey: PropertyValue
-  value: BehavioralEvents
 - schemaKey: PropertyValue
   value: Units
 wasAttributedTo:
@@ -64,16 +43,22 @@
     schemaKey: SexType
   species:
     identifier: http://purl.obolibrary.org/obo/NCBITaxon_10090
-    name: House mouse
+    name: Mus musculus - House mouse
     schemaKey: SpeciesType
 wasGeneratedBy:
+- description: ''
+  name: Acquisition session
+  schemaKey: Session
+  startDate: '2017-03-10T00:00:00-06:00'
 - description: Metadata generated by DANDI cli
-  id: urn:uuid:5267ee13-ae30-4466-aad4-4822019df331
+  endDate: '2023-03-03T11:14:56.001978-05:00'
+  id: urn:uuid:3d81857a-ac93-4a3f-9a11-867c084731c5
   name: Metadata generation
   schemaKey: Activity
+  startDate: '2023-03-03T11:14:47.600744-05:00'
   wasAssociatedWith:
   - identifier: RRID:SCR_019009
     name: DANDI Command Line Interface
     schemaKey: Software
     url: https://github.com/dandi/dandi-cli
-    version: 0.22.0
+    version: 0.50.1+4.gdc68e02e

2023-03-03 11:14:56,110 [    INFO] Saving new asset metadata
2023-03-03 11:14:56,460 [    INFO] Logs saved in /home/yoh/.cache/dandi-cli/log/20230303161444Z-3510157.log
dandi service-scripts reextract-metadata --diff --when newer-schema-version   4.94s user 2.50s system 57% cpu 13.018 total
```
</details>